odl/discr/lp_discr.py

```diff
@@ -434,15 +409,103 @@ def _dist(self, x, y):
         if self.is_uniformly_weighted:
             return super()._dist(x, y)
         else:
-            # TODO: implement without copying x
+            # TODO: implement without copying x and y
             bdry_fracs = self.partition.boundary_cell_fractions
             func_list = _scaling_func_list(bdry_fracs, exponent=self.exponent)
-
             arrs = [apply_on_boundary(vec, func=func_list, only_once=False)
                     for vec in (x, y)]
-
             return super()._dist(self.element(arrs[0]), self.element(arrs[1]))
 
+    # TODO: add byaxis_out when discretized tensor-valued functions are
+    # available
+
+    @property
+    def byaxis_in(self):
+        """Object to index along input (domain) dimensions.
+
+        Examples
+        --------
+        Indexing with integers or slices:
+
+        >>> space = odl.uniform_discr([0, 0, 0], [1, 2, 3], (5, 10, 15))
+        >>> space.byaxis_in[0]
+        uniform_discr(0.0, 1.0, 5)
+        >>> space.byaxis_in[1]
+        uniform_discr(0.0, 2.0, 10)
+        >>> space.byaxis_in[1:]
+        uniform_discr([0.0, 0.0], [2.0, 3.0], (10, 15))
+
+        Lists can be used to stack spaces arbitrarily:
+
+        >>> space.byaxis_in[[2, 1, 2]]
+        uniform_discr([0.0, 0.0, 0.0], [3.0, 2.0, 3.0], (15, 10, 15))
+        """
+        space = self
+
+        class DiscreteLpByaxisIn(object):
+
+            """Helper class for indexing by domain axes."""
+
+            def __getitem__(self, indices):
+                """Return ``self[indices]``.
+
+                Parameters
+                ----------
+                indices : index expression
+                    Object used to index the space domain.
+
+                Returns
+                -------
+                space : `DiscreteLp`
+                    The resulting space with indexed domain and otherwise
+                    same properties (except possibly weighting).
+                """
+                fspace = space.uspace.byaxis_in[indices]
+                part = space.partition.byaxis[indices]
+
+                if isinstance(space.weighting, NoWeighting):
+                    # Extra case since NoWeighting is a subclass of
+                    # ConstWeighting
```

Comment:
Do we want to keep the singleton pattern for the `const=1` case? Does it matter in terms of efficiency?

Reply:
Related issue: This change will also wipe the `'none'` option to weighting in `uniform_discr`. I suggest we change that parameter completely to default to `None` and propagate as-is to the `dspace` constructor. That way we can easily construct arbitrarily weighted spaces.

Reply:
Thought some more about this and second thought we should NOT remove `NoWeighting`. Specifically it conveys a different intention and distinguishes "accidental" no-weighting from intentional. That is, `uniform_discr(0, 5, 5)` has a weighting that is "conceptually" different from `rn(5)`, even if it practically is the same. Right?

Reply:
Well, they would still be distinguishable by `type()`. A more tricky case would be `uniform_discr([0, 0], [4, 4], (4, 4))` vs. `uniform_discr([0, 0], [4, 4], (2, 8))`. Both would result in constant weighting with constant 1.0. Hence not weighted. Basically the `space.is_weighted` would reduce to a pure convenience, or to something that allows a few optimizations.

Would anything go wrong? Since implementation-wise, also the old "no weighting" case was equivalent to constant 1, so I expect trouble only when doing things to axes, like `reduce` or `outer`, or indexing `byaxis_in`.
But I think everything is fine, since `DiscreteLp` always knows how the constant came to be, so it can distinguish between "arbitrary" constants and cell volume. For the latter, it will do the reasonable thing and re-compute the constant.

As a matter of fact, I already made most of the changes, and only one test fails that was constructed to test weighting propagation in a way that has to fail now.

Random thought: when we go about implementing lists of weights (like arrays per axis with broadcasting), we can obviously also allow constants per axis, and as an optimization multiply them together, making it equivalent to a single constant. The good thing would be that `TensorSpace` would know how to handle adding/removing axes in terms of weighting.